kingpin/actors/rightscale/alerts.py

```diff
@@ -0,0 +1,364 @@
+        'threshold': (str, REQUIRED,
+                      'The threshold of the condition sentence.'),
+        'variable': (str, REQUIRED,
+                     'The RRD variable of the condition sentence.'),
+        'vote_tag': (str, None,
+                     ('Should correspond to a vote tag on a ServerArray if ',
+                      'vote to grow or shrink.')),
+        'vote_type': (str, None,
+                      ('Vote to grow or shrink a ServerArray when the alert ',
+                       'is triggered. Must either escalate or vote.'))
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Validate the user-supplied parameters at instantiation time."""
+        super(Create, self).__init__(*args, **kwargs)
+        # By default, we're strict on our array array validation
+        self._array_raise_on = 'notfound'
+        self._array_allow_mock = False
+
+        if not self.option('strict_array'):
+            self._array_raise_on = None
+            self._array_allow_mock = True
+
+    @gen.coroutine
+    def _execute(self):
+        # Find the array we're adding an alert spec to. Specifically, we need
+        # the servers HREF.
+        array = yield self._find_server_arrays(
+            self.option('array'),
+            raise_on=self._array_raise_on,
+            allow_mock=self._array_allow_mock)
+        self.log.info('Found %s (%s)' % (array.soul['name'], array.href))
+
+        # Add all of the required parameters to a dictionary
+        params = {
+            'condition': self.option('condition'),
+            'description': self.option('description'),
+            'duration': self.option('duration'),
+            'file': self.option('file'),
+            'name': self.option('name'),
+            'subject_href': array.href,
+            'threshold': self.option('threshold'),
+            'variable': self.option('variable'),
+        }
+
+        # Generate the RightScale parameters that we need to pass in when
+        # creating the alert. The optional parameters should not be passed in
+        # if their option value came in as None.
+        _optional_params = [
+            'description', 'escalation_name', 'vote_tag', 'vote_type'
+        ]
+        for optional in _optional_params:
+            if self.option(optional):
+                params[optional] = self.option(optional)
+
+        params = self._generate_rightscale_params('alert_spec', params)
+        self.log.debug('Generated params: %s' % params)
+
+        if self._dry:
+            # In dry run mode, just log out what we would have done.
+            self.log.info('Would have created the alert spec \"%s\" on %s' %
+                          (self.option('name'), array.soul['name']))
+        else:
+            # We're really doin this. If we get a known exception back, handle
+            # it. Otherwise, raise it.
+            try:
+                yield self._client.create_resource(
```

Comment:
This should be split into its own function, wrapped in `@retry` -- so that HTTPError is handled automatically